spring-integration-core/src/main/java/org/springframework/integration/metadata/MetadataStoreReaper.java

```diff
@@ -0,0 +1,175 @@
+/*
+ * Copyright 2018 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.integration.metadata;
+
+import java.time.Duration;
+import java.time.Instant;
+import java.time.LocalDateTime;
+import java.time.ZonedDateTime;
+import java.util.TimeZone;
+import java.util.concurrent.locks.ReentrantLock;
+import java.util.function.Predicate;
+
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+
+import org.springframework.context.SmartLifecycle;
+import org.springframework.util.Assert;
+
+/**
+ * {@link Runnable} component that deletes entries from a {@link MetadataStore}.
+ * By default, it expects the value of each entry to be a timestamp, as is the default configuration when using the
+ * {@link org.springframework.integration.handler.advice.IdempotentReceiverInterceptor} and the
+ * {@link org.springframework.integration.selector.MetadataStoreSelector}. You can inject a
+ * {@link Predicate} to implement an expiration condition.
+ *
+ * @author David Turanski
+ **/
+
+public class MetadataStoreReaper implements Runnable, SmartLifecycle {
+
+	private final static Log logger = LogFactory.getLog(MetadataStoreReaper.class);
+
+	private final MetadataStore metadataStore;
+
+	private final Predicate<String> isExpired;
+
+	private volatile boolean running;
+
+	private final ReentrantLock lifecycleLock = new ReentrantLock();
+
+	private volatile int phase = 0;
+
+	private volatile boolean autoStartup = true;
+
+	/**
+	 * Remove entries from a {@link MetadataStore} based on a maximum time-to-live.
+	 * @param metadataStore the MetadataStore.
+	 * @param timeToLive the maximum time-to-live.
+	 */
+	public MetadataStoreReaper(MetadataStore metadataStore, Duration timeToLive) {
+		this(metadataStore, value -> {
+			try {
+				long timestamp = Long.valueOf(value);
```

Comment:
I would say that this is too custom solution and should not be as a part of the Framework.

Can we keep this `MetadataStoreReaper` in your target project?
That's just a bad luck that your use-case is based on the fact that we store a `timestamp` in the value, but there are many other scenarios when it is not true. 

So, 👎 from me to merge this into `master`.

Reply:
IMHO, The reaper is a good thing to have in SI and doesn't hurt anything. People don't have to use it.  If they choose to use it, the value should include date-time info. But to your point, I can will remove this constructor and function and implement it at the app level where it makes sense. The reaper should be a common component though, since it addresses a core problem with the metadataStore.   

A more comprehensive approach I think would require the implementation to support the clean up, similar to the way `MessageGroupReaper` delegates to the `MessageGroupStore`.  All `MetadataStore`s would have to implement the clean up. That's the only way I can see to have a universal solution without violating encapsulation.  

My suggestion is to keep this and refactor for a more comprehensive approach if there is demand.

Reply:
I just saw your comment below:  `... I would say that it is a time to add expireEntries(Predicate<String>) into the MetadataStore contract!`